File: platform/polyglot/gemini_openai_sdk/py/gemini_openai_sdk/devtools.py
"""
DevTools Module - Developer Utilities

Provides debugging utilities for SDK development and testing.
"""

import json
from typing import Any, Dict, List, Optional

from env_resolver import resolve_gemini_env

from .constants import CHAT_ENDPOINT, DEFAULT_MODEL, DEFAULTS, MODELS, SYSTEM_PROMPT
from .gemini_client import GeminiClient
from .helpers import get_api_key, get_headers, get_model
from .logger import create

logger = create("gemini_openai_sdk", __file__)

_gemini_env = resolve_gemini_env()


def create_debug_client(
    model: str = DEFAULT_MODEL,
    verbose: bool = True,
) -> GeminiClient:
    """
    Create a client with verbose logging enabled.

    Args:
        model: Model type to use
        verbose: Whether to enable verbose logging (default: True)

    Returns:
        GeminiClient configured for debugging
    """
    import os

    if verbose:
        os.environ["LOG_LEVEL"] = "DEBUG"

    logger.info("create_debug_client: creating debug client", model=model)
    return GeminiClient(model=model)


def inspect_request(
    messages: List[Dict[str, str]],
    model: Optional[str] = None,
    temperature: Optional[float] = None,
    max_tokens: Optional[int] = None,
    **kwargs: Any,
) -> Dict[str, Any]:
    """
    Inspect what would be sent to API without executing.

    Args:
        messages: Messages array
        model: Model type
        temperature: Temperature setting
        max_tokens: Max tokens setting
        **kwargs: Additional parameters

    Returns:
        Dictionary showing request details
    """
    logger.debug("inspect_request: building request preview")

    api_key = _gemini_env.api_key
    resolved_model = get_model(model or DEFAULT_MODEL)

    payload = {
        "model": resolved_model,
        "messages": messages,
        "temperature": temperature if temperature is not None else DEFAULTS["temperature"],
        "max_tokens": max_tokens if max_tokens is not None else DEFAULTS["max_tokens"],
        **kwargs,
    }

    headers = get_headers(api_key or "API_KEY_NOT_SET")
    # Mask auth header for display
    headers_display = {
        **headers,
        "Authorization": "Bearer ***" + (api_key[-4:] if api_key else "****"),
    }

    return {
        "endpoint": CHAT_ENDPOINT,
        "method": "POST",
        "headers": headers_display,
        "payload": payload,
        "api_key_set": bool(api_key),
        "resolved_model": resolved_model,
    }


def mock_response(
    content: str = "Mock response content",
    model: str = "gemini-2.0-flash",
    finish_reason: str = "stop",
    prompt_tokens: int = 10,
    completion_tokens: int = 20,
) -> Dict[str, Any]:
    """
    Create a mock API response for testing.

    Args:
        content: Response content
        model: Model name in response
        finish_reason: Finish reason
        prompt_tokens: Token count
        completion_tokens: Token count

    Returns:
        Mock API response dictionary
    """
    logger.debug("mock_response: creating mock response")

    return {
        "id": "chatcmpl-mock-12345",
        "object": "chat.completion",
        "created": 1234567890,
        "model": model,
        "choices": [
            {
                "index": 0,
                "message": {
                    "role": "assistant",
                    "content": content,
                },
                "finish_reason": finish_reason,
            }
        ],
        "usage": {
            "prompt_tokens": prompt_tokens,
            "completion_tokens": completion_tokens,
            "total_tokens": prompt_tokens + completion_tokens,
        },
    }


def mock_stream_chunks(
    content: str = "Hello world",
    model: str = "gemini-2.0-flash",
) -> List[str]:
    """
    Create mock streaming chunks for testing.

    Args:
        content: Content to split into chunks
        model: Model name

    Returns:
        List of SSE data strings
    """
    logger.debug("mock_stream_chunks: creating mock stream")

    chunks = []

    # First chunk with role
    chunks.append(json.dumps({
        "id": "chatcmpl-mock-stream",
        "object": "chat.completion.chunk",
        "model": model,
        "choices": [
            {
                "index": 0,
                "delta": {"role": "assistant"},
                "finish_reason": None,
            }
        ],
    }))

    # Content chunks
    for char in content:
        chunks.append(json.dumps({
            "id": "chatcmpl-mock-stream",
            "object": "chat.completion.chunk",
            "model": model,
            "choices": [
                {
                    "index": 0,
                    "delta": {"content": char},
                    "finish_reason": None,
                }
            ],
        }))

    # Final chunk
    chunks.append(json.dumps({
        "id": "chatcmpl-mock-stream",
        "object": "chat.completion.chunk",
        "model": model,
        "choices": [
            {
                "index": 0,
                "delta": {},
                "finish_reason": "stop",
            }
        ],
    }))

    return chunks


def show_config() -> Dict[str, Any]:
    """
    Display current SDK configuration.

    Returns:
        Configuration dictionary
    """
    api_key = _gemini_env.api_key

    return {
        "models": MODELS,
        "default_model": DEFAULT_MODEL,
        "base_url": CHAT_ENDPOINT.rsplit("/", 1)[0],
        "chat_endpoint": CHAT_ENDPOINT,
        "defaults": DEFAULTS,
        "system_prompt": SYSTEM_PROMPT[:50] + "..." if len(SYSTEM_PROMPT) > 50 else SYSTEM_PROMPT,
        "api_key_configured": bool(api_key),
        "api_key_preview": "***" + api_key[-4:] if api_key else None,
    }


def validate_environment() -> Dict[str, Any]:
    """
    Validate environment setup.

    Returns:
        Validation result with any issues found
    """
    issues = []

    api_key = _gemini_env.api_key
    if not api_key:
        issues.append("GEMINI_API_KEY environment variable not set")

    import sys
    python_version = sys.version_info
    if python_version < (3, 11):
        issues.append(f"Python 3.11+ required, found {python_version.major}.{python_version.minor}")

    try:
        import httpx
    except ImportError:
        issues.append("httpx package not installed")

    return {
        "valid": len(issues) == 0,
        "issues": issues,
        "python_version": f"{python_version.major}.{python_version.minor}.{python_version.micro}",
        "api_key_set": bool(api_key),
    }

File: platform/polyglot/gemini_openai_sdk/mjs/gemini-openai-sdk/devtools.mjs
/**
 * DevTools Module - Developer Utilities
 *
 * Provides debugging utilities for SDK development and testing.
 */

import { create } from './logger.mjs';
import { MODELS, DEFAULT_MODEL, CHAT_ENDPOINT, DEFAULTS, SYSTEM_PROMPT } from './constants.mjs';
import { getApiKey, getModel, getHeaders } from './helpers.mjs';
import { GeminiClient } from './gemini-client.mjs';
import { resolveGeminiEnv } from '@internal/env-resolver';

const logger = create('gemini-openai-sdk', import.meta.url);

const _geminiEnv = resolveGeminiEnv();

/**
 * Create a client with verbose logging enabled.
 * @param {string} [model] - Model type
 * @param {boolean} [verbose=true] - Enable verbose logging
 * @returns {GeminiClient} Debug-configured client
 */
export function createDebugClient(model = DEFAULT_MODEL, verbose = true) {
  if (verbose) {
    process.env.LOG_LEVEL = 'DEBUG';
  }

  logger.info('createDebugClient: creating debug client', { model });
  return new GeminiClient({ model });
}

/**
 * Inspect what would be sent to API without executing.
 * @param {Array} messages - Messages array
 * @param {object} [options] - Request options
 * @returns {object} Request preview
 */
export function inspectRequest(messages, options = {}) {
  logger.debug('inspectRequest: building request preview');

  const apiKey = _geminiEnv.apiKey || null;
  const resolvedModel = getModel(options.model || DEFAULT_MODEL);

  const payload = {
    model: resolvedModel,
    messages,
    temperature: options.temperature ?? DEFAULTS.temperature,
    max_tokens: options.max_tokens ?? DEFAULTS.max_tokens,
    ...options,
  };

  delete payload.model; // Remove from spread
  payload.model = resolvedModel; // Re-add at top

  const headers = getHeaders(apiKey || 'API_KEY_NOT_SET');
  const headersDisplay = {
    ...headers,
    Authorization: 'Bearer ***' + (apiKey ? apiKey.slice(-4) : '****'),
  };

  return {
    endpoint: CHAT_ENDPOINT,
    method: 'POST',
    headers: headersDisplay,
    payload,
    api_key_set: !!apiKey,
    resolved_model: resolvedModel,
  };
}

/**
 * Create a mock API response for testing.
 * @param {object} [options] - Mock options
 * @returns {object} Mock API response
 */
export function mockResponse(options = {}) {
  logger.debug('mockResponse: creating mock response');

  const {
    content = 'Mock response content',
    model = 'gemini-2.0-flash',
    finishReason = 'stop',
    promptTokens = 10,
    completionTokens = 20,
  } = options;

  return {
    id: 'chatcmpl-mock-12345',
    object: 'chat.completion',
    created: Math.floor(Date.now() / 1000),
    model,
    choices: [
      {
        index: 0,
        message: {
          role: 'assistant',
          content,
        },
        finish_reason: finishReason,
      },
    ],
    usage: {
      prompt_tokens: promptTokens,
      completion_tokens: completionTokens,
      total_tokens: promptTokens + completionTokens,
    },
  };
}

/**
 * Create mock streaming chunks for testing.
 * @param {string} [content='Hello world'] - Content to split
 * @param {string} [model='gemini-2.0-flash'] - Model name
 * @returns {Array<string>} SSE data strings
 */
export function mockStreamChunks(content = 'Hello world', model = 'gemini-2.0-flash') {
  logger.debug('mockStreamChunks: creating mock stream');

  const chunks = [];

  // First chunk with role
  chunks.push(JSON.stringify({
    id: 'chatcmpl-mock-stream',
    object: 'chat.completion.chunk',
    model,
    choices: [
      {
        index: 0,
        delta: { role: 'assistant' },
        finish_reason: null,
      },
    ],
  }));

  // Content chunks
  for (const char of content) {
    chunks.push(JSON.stringify({
      id: 'chatcmpl-mock-stream',
      object: 'chat.completion.chunk',
      model,
      choices: [
        {
          index: 0,
          delta: { content: char },
          finish_reason: null,
        },
      ],
    }));
  }

  // Final chunk
  chunks.push(JSON.stringify({
    id: 'chatcmpl-mock-stream',
    object: 'chat.completion.chunk',
    model,
    choices: [
      {
        index: 0,
        delta: {},
        finish_reason: 'stop',
      },
    ],
  }));

  return chunks;
}

/**
 * Display current SDK configuration.
 * @returns {object} Configuration object
 */
export function showConfig() {
  const apiKey = _geminiEnv.apiKey || null;

  return {
    models: MODELS,
    default_model: DEFAULT_MODEL,
    base_url: CHAT_ENDPOINT.split('/').slice(0, -1).join('/'),
    chat_endpoint: CHAT_ENDPOINT,
    defaults: DEFAULTS,
    system_prompt: SYSTEM_PROMPT.length > 50
      ? SYSTEM_PROMPT.slice(0, 50) + '...'
      : SYSTEM_PROMPT,
    api_key_configured: !!apiKey,
    api_key_preview: apiKey ? '***' + apiKey.slice(-4) : null,
  };
}

/**
 * Validate environment setup.
 * @returns {Promise<object>} Validation result
 */
export async function validateEnvironment() {
  const issues = [];

  const apiKey = _geminiEnv.apiKey || null;
  if (!apiKey) {
    issues.push('GEMINI_API_KEY environment variable not set');
  }

  const nodeVersion = process.versions.node.split('.').map(Number);
  if (nodeVersion[0] < 20) {
    issues.push(`Node.js 20.x+ required, found ${process.version}`);
  }

  // Check for undici
  try {
    await import('undici');
  } catch {
    issues.push('undici package not installed');
  }

  return {
    valid: issues.length === 0,
    issues,
    node_version: process.version,
    api_key_set: !!apiKey,
  };
}

export default {
  createDebugClient,
  inspectRequest,
  mockResponse,
  mockStreamChunks,
  showConfig,
  validateEnvironment,
};
